### PR TITLE
allow more workers for ILM expiration

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
allow more workers for ILM expiration

## Motivation and Context
- add configurable workers for ILM expiration
- default to 50% of MAXPROCS
- also make sure to have a larger queue for 100k 
  instead of 10k, we are currently under a lot of 
  backlogs if there are many objects to expire

## How to test this PR?
Needs a valid ILM configuration with NewerNonCurrent, 
as well as regular expiry. However, this is mostly 
tested with expiration to trigger faster and dequeue 
faster when there is a large backlog.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/bucket-lifecycle.go [hunk 1]
```diff
@@ -116,26 +120,49 @@ var globalExpiryState *expiryState
...
+	}
+
 	go func() {
 		for t := range globalExpiryState.byDaysCh {
-			if t.objInfo.TransitionedObject.Status != "" {
-				applyExpiryOnTransitionedObject(ctx, objectAPI, t.objInfo, t.restoredObject)
-			} else {
-				applyExpiryOnNonTransitionedObjects(ctx, objectAPI, t.objInfo, t.versionExpiry)
-			}
+			ewk.Give()
```
Comment (klauspost) [inline cmd/bucket-lifecycle.go]:
```suggestion
			ewk.Take()
```

File: cmd/bucket-lifecycle.go [hunk 2]
```diff
@@ -116,26 +120,49 @@ var globalExpiryState *expiryState
...
 	go func() {
 		for t := range globalExpiryState.byNewerNoncurrentCh {
-			deleteObjectVersions(ctx, objectAPI, t.bucket, t.versions)
+			nwk.Take()
+			go func(t newerNoncurrentTask) {
+				defer nwk.Give()
+				deleteObjectVersions(ctx, objectAPI, t.bucket, t.versions)
+			}(t)
 		}
+		nwk.Wait()
```
Comment (klauspost) [inline cmd/bucket-lifecycle.go:165]: Is there a point in waiting here? It just exits the goroutine anyway.

```suggestion
```

File: cmd/bucket-lifecycle.go [hunk 3]
```diff
@@ -116,26 +120,49 @@ var globalExpiryState *expiryState
...
+			go func(t expiryTask) {
+				defer ewk.Give()
+				if t.objInfo.TransitionedObject.Status != "" {
+					applyExpiryOnTransitionedObject(ctx, objectAPI, t.objInfo, t.restoredObject)
+				} else {
+					applyExpiryOnNonTransitionedObjects(ctx, objectAPI, t.objInfo, t.versionExpiry)
+				}
+			}(t)
 		}
+		ewk.Wait()
```
Comment (klauspost) [inline cmd/bucket-lifecycle.go:155]:
```suggestion
```

File: cmd/bucket-lifecycle.go [hunk 4]
```diff
@@ -116,26 +120,49 @@ var globalExpiryState *expiryState
...
-		byNewerNoncurrentCh: make(chan newerNoncurrentTask, 10000),
+		byDaysCh:            make(chan expiryTask, 100000),
+		byNewerNoncurrentCh: make(chan newerNoncurrentTask, 100000),
 	}
 }
 
 func initBackgroundExpiry(ctx context.Context, objectAPI ObjectLayer) {
 	globalExpiryState = newExpiryState()
+
+	workerSize, _ := strconv.Atoi(env.Get("_MINIO_ILM_EXPIRY_WORKERS", strconv.Itoa(runtime.GOMAXPROCS(0)/2)))
```
Comment (klauspost) [inline cmd/bucket-lifecycle.go]: Don't risk 0 workers (whatever that does):
```suggestion
	workerSize, _ := strconv.Atoi(env.Get("_MINIO_ILM_EXPIRY_WORKERS", strconv.Itoa((runtime.GOMAXPROCS(0)+1)/2)))
```

**Review by klauspost — CHANGES_REQUESTED** [review]

File: cmd/bucket-lifecycle.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/bucket-lifecycle.go:165]: well, want to wait simply via pattern the channel would never close anyways, even if it does we simply wait for the ILM to complete before we give up the goroutine.

Just wanted to make sure that any future movements in the code this code is protected.

File: cmd/bucket-lifecycle.go [hunk 5]
```diff
@@ -116,26 +120,49 @@ var globalExpiryState *expiryState
...
-		byNewerNoncurrentCh: make(chan newerNoncurrentTask, 10000),
+		byDaysCh:            make(chan expiryTask, 100000),
+		byNewerNoncurrentCh: make(chan newerNoncurrentTask, 100000),
 	}
 }
 
 func initBackgroundExpiry(ctx context.Context, objectAPI ObjectLayer) {
 	globalExpiryState = newExpiryState()
+
+	workerSize, _ := strconv.Atoi(env.Get("_MINIO_ILM_EXPIRY_WORKERS", strconv.Itoa(runtime.GOMAXPROCS(0)+1/2)))
```
Comment (klauspost) [inline cmd/bucket-lifecycle.go]: Check operator precedence. `runtime.GOMAXPROCS(0)+1/2` is `runtime.GOMAXPROCS(0)+(1/2)`.

**Review by klauspost — APPROVED** [review]

**Review by poornas — APPROVED** [review]

**Review by krisis — APPROVED** [review]

**Pull request merged**